src/darwin/Framework/CHIPTests/MTRPerControllerStorageTests.m

```diff
@@ -2172,10 +2172,944 @@ - (void)doTestSubscriptionPoolWithSize:(NSInteger)subscriptionPoolSize
+        MTRAttributePathKey : [MTRAttributePath attributePathWithEndpointID:testEndpoint clusterID:@(MTRClusterIDTypeDescriptorID) attributeID:@(MTRAttributeIDTypeClusterDescriptorAttributeServerListID)],
+        MTRDataKey : @ {
+            MTRDataVersionKey : dataVersionForServerList,
+            MTRTypeKey : MTRArrayValueType,
+            MTRValueKey : testClusterDataValue,
+        }
+    } ];
+
+    XCTestExpectation * attributeDataReceivedExpectation = [self expectationWithDescription:@"Injected Attribute data received"];
+    XCTestExpectation * reportEndExpectation = [self expectationWithDescription:@"Injected Attribute data report ended"];
+    delegate.onAttributeDataReceived = ^(NSArray<NSDictionary<NSString *, id> *> * attributeReport) {
+        attributeReportsReceived += attributeReport.count;
+        XCTAssert(attributeReportsReceived > 0);
+        [attributeDataReceivedExpectation fulfill];
+    };
+
+    delegate.onReportEnd = ^{
+        // Verify that the data store is populated with cluster index and cluster data for endpoints 0, 1 and 2.
+        // Also _persistedClusters and _persistedClusterData has the cluster paths for all clusters in the cluster index. Nothing is deleted.
+        dispatch_sync(self->_storageQueue, ^{
+            XCTAssertTrue([[controller.controllerDataStore _fetchEndpointIndexForNodeID:deviceID] isEqualToArray:testEndpoints]);
+            for (NSNumber * endpoint in testEndpoints) {
+                XCTAssertNotNil(initialClusterIndex);
+                for (NSNumber * cluster in [initialClusterIndex objectAtIndex:endpoint.unsignedShortValue]) {
+                    XCTAssertNotNil([controller.controllerDataStore _fetchClusterDataForNodeID:deviceID endpointID:endpoint clusterID:cluster]);
+                    MTRClusterPath * path = [MTRClusterPath clusterPathWithEndpointID:endpoint clusterID:cluster];
+                    XCTAssertTrue([device _persistedClusterContains:path]);
+                    XCTAssertNotNil([device _getPersistedClusterDataForPath:path]);
+                }
+            }
+        });
+
+        [reportEndExpectation fulfill];
+    };
+
+    [device unitTestInjectAttributeReport:attributeReport fromSubscription:YES];
+
+    [self waitForExpectations:@[ attributeDataReceivedExpectation, reportEndExpectation ] timeout:60];
+
+    [controller.controllerDataStore clearAllStoredClusterData];
+    NSDictionary * storedClusterDataAfterClear = [controller.controllerDataStore getStoredClusterDataForNodeID:deviceID];
+    XCTAssertEqual(storedClusterDataAfterClear.count, 0);
+
+    [controller removeDevice:device];
+    // Reset our commissionee.
+    __auto_type * baseDevice = [MTRBaseDevice deviceWithNodeID:deviceID controller:controller];
+    ResetCommissionee(baseDevice, queue, self, kTimeoutInSeconds);
+
+    [controller shutdown];
+    XCTAssertFalse([controller isRunning]);
+}
+
+- (void)test016_TestDataStoreMTRDevicePruneOrphanedClusters
```

Comment:
Nothing here is orphaned.

See comments above about factoring things out here so that we don't have so many copies of the same code that then have to be kept in sync.

Reply:
fixed